src/sage/tensor/modules/comp_numpy.py

```diff
@@ -0,0 +1,1110 @@
+          ``args = [:]`` (``slice(None)``)
+
+        EXAMPLES::
+
+            sage: from sage.tensor.modules.comp_numpy import ComponentNumpy
+            sage: c = ComponentNumpy(ZZ, [1,2,3], 2)
+            sage: c.__setitem__((0,1), -4)
+            sage: c[:]
+            array([[ 0., -4.,  0.],
+                   [ 0.,  0.,  0.],
+                   [ 0.,  0.,  0.]])
+            sage: c[0,1] = -4
+            sage: c[:]
+            array([[ 0., -4.,  0.],
+                   [ 0.,  0.,  0.],
+                   [ 0.,  0.,  0.]])
+            sage: c.__setitem__(slice(None), [[0, 1, 2], [3, 4, 5], [6, 7, 8]])
+            sage: c[:]
+            array([[0., 1., 2.],
+                  [3., 4., 5.],
+                  [6., 7., 8.]])
+            sage: c = ComponentNumpy(ZZ, [1,2,3], 2, start_index=[1,2])
+            sage: c.__setitem__(slice(None), [[0, 1, 2], [3, 4, 5], [6, 7, 8]])
+            sage: c[1:4:2] = [[2, 1, 0], [5, 4, 3]]
+            sage: c[:]
+            array([[2., 1., 0.],
+                   [3., 4., 5.],
+                   [5., 4., 3.]])
+            sage: c[2:5] = [[0, 1, 2], [3, 4, 5], [6, 7, 8]]
+            Traceback (most recent call last):
+            ...
+            IndexError: [start:stop] not in range [1,4]
+        """
+        if isinstance(args, list):  # case of [[...]] syntax
+            if isinstance(args[0], slice):
+                indices = args[0]
+            elif isinstance(args[0], (tuple, list)): # to ensure equivalence between
+                indices = args[0]           # [[(i,j,...)]] or [[[i,j,...]]] and [[i,j,...]]
+            else:
+                indices = tuple(args)
+        else:
+            # Determining from the input the list of indices and the format
+            if isinstance(args, (int, Integer, slice)):
+                indices = args
+            elif isinstance(args[0], slice):
+                indices = args[0]
+            elif len(args) == self._nid:
+                indices = args
+
+        if isinstance(indices, slice):
+            start, stop = indices.start, indices.stop
+            range_start, range_end = self._sindex[0], self._sindex[0] + self._shape[0]
+
+            if indices.start is not None:
+                start = indices.start - range_start
+            else:
+                start = 0
+            if indices.stop is not None:
+                stop = indices.stop - range_start
+            else:
+                stop = range_end - range_start
+            if not ((0 <= start <= range_end - range_start - 1) and
+                    (0 <= stop <= range_end - range_start)):
+                raise IndexError("[start:stop] not in range [{},{}]"
+                    .format(range_start, range_end))
+
+            self._comp[start:stop:indices.step] = value
+        else:
+            ind = self._check_indices(indices)
+            self._comp[ind] = value
+
+    def _broadcast(self, other):
+        r"""
+        broadcast self with other
```

Comment:
Best to include an explanation (or link to explanation) of this jargon, which is so far not used in Sage